client_test.go

```diff
@@ -208,3 +212,24 @@ func TestUseFstatChecked(t *testing.T) {
 	testFstatOption(t, UseFstat(true), true)
 	testFstatOption(t, UseFstat(false), false)
 }
+
+type sink struct{}
+
+func (*sink) Close() error                { return nil }
+func (*sink) Write(p []byte) (int, error) { return len(p), nil }
+
+func TestBrokenServer(t *testing.T) {
+	// Simulate a server that sends a single packet with too small a length,
+	// regardless of what the client sends.
+
+	var packet [4]byte
+	w := &sink{}
```

Comment:
Changing NewClientPipe would, strictly speaking, be a backward incompatible change. Also, it suggests that the caller should close the pipe, leading to double Close errors in client code.

Reply:
No, it would not be a backwards-incompatible change. All `io.WriteCloser` are also `io.Writer`, and with `if closer, ok := writer.(io.Closer); ok { closer.Close() }` it also retains the same behavior. This also is not unusual design: https://golang.org/pkg/crypto/cipher/#StreamWriter already exhibits this same design for `io.Writer`, and https://golang.org/pkg/net/http/#Client.Post uses it for `io.Reader`.

Since this is not a reciever method, there are no interface mismatch concerns, so really the only “incompatible” thing about it is if someone explicitly uses the type: `func(io.Reader, io.WriteCloser, ...sftp.ClientOption) (*sftp.Client, error)`,  but avoiding the type in declarations (`var newClientPipe = sftp.NewClientPipe`) should be proper hygiene, so that the type is auto-updated, leaving really only a situation where the function itself it is either a parameter to another function, or a field in a struct. Both of those seem quite unlikely, and unnecessary.

Reply:
It should also add, this is a convenience for testing more than a good API fix, and it’s not without _any_ risk, it just should be low risk. But we should perhaps tackle it all on its own.